core/peerset/src/lib.rs

```diff
@@ -236,68 +351,85 @@ impl PeersetMut {
 	///
 	/// Because of concurrency issues, it is acceptable to call `incoming` with a `PeerId` the
 	/// peerset is already connected to, in which case it must not answer.
-	pub fn incoming(&self, peer_id: PeerId, index: IncomingIndex) {
-		let mut inner = self.parent.inner.lock();
-		if inner.out_slots.iter().chain(inner.in_slots.iter()).any(|s| s.as_ref() == Some(&peer_id)) {
-			return
+	pub fn incoming(&mut self, peer_id: PeerId, index: IncomingIndex) {
+		// if `reserved_only` is set, but this peer is not a part of our discovered list,
+		// a) it is not reserved, so we reject the connection
+		// b) we are already connected to it, so we reject the connection
+		if self.data.reserved_only && !self.data.discovered.is_reserved(&peer_id) {
+			self.message_queue.push_back(Message::Reject(index));
+			return;
+		}
+
+		// check if we are already connected to this peer
+		if self.data.out_slots.contains(&peer_id) {
+			self.message_queue.push_back(Message::Reject(index));
```

Comment:
Must **not** send `Reject`, see the doc-comment of the function.

Reply:
Thanks for pointing this out. May I accept why is it designed like that?

Reply:
I don't remember, to be honest. There was a long debate about this and this was the outcome.